DOCS: Add deployment docs over GCP and Azure (#103)
moja-global/FLINT.Cloud: docs/deployment/index.rst

.. _Deployment:

FLINT.Cloud Deployment
####################

This section guides you through the process of deploying FLINT.Cloud to
your own server, through a public cloud service provider. We are using
Google Cloud Platform (GCP) and Azure, as of now, to deploy FLINT.Cloud
APIs to be consumed for your FLINT-specific applications, like FLINT-UI.
This guide assumes that you have some preliminary understanding of the
concepts of GCP and Azure and have deployed an application before.

Google Cloud Platform
---------------------

To deploy FLINT.Cloud to Google Cloud Platform, we provide a
production-grade setup with a Layered Architecture Setup on top of the
Google Cloud. This setup includes the usage of `Terraform`_, an
Infrastructure-as-a-Code tool, to deploy the infrastructure and the
application to be used.

To deploy the FLINT.Cloud to Google Cloud Platform, follow these steps:

> 1. Create a GCP service account with project owner permissions in your

Reviewer: Dont recommend giving project owner permission for service accounts. Give the permission required to deploy the infrastructure such as Compute admin etc.

   project. This is used by Terraform to provision all the necessary
   resources.
2. Copy ``main.tf`` from the ``layered`` directory of this repository to
   your Cloud Console machine.
3. In ``main.tf``, change the project variable to your ``project ID``.

>    Change any other variables if necessary.

Reviewer 2: ```suggestion
   Change any other variables, if necessary.
```

> 4. Download the key for the service account created in the ``Step 1``
>    (in JSON format) to your project's Cloud Console machine. Rename it

Reviewer 2: ```suggestion
4. Download the key in JSON format for the service account created in ``Step 1``
to your project's Cloud Console machine. Rename it
```

>    as ``service_account.json``.

Reviewer 2: ```suggestion
   to ``service_account.json``.
```

5. Run ``terraform apply``. After this command finishes, it should

>    output the URL to FLINT Cloud (ingress).

Reviewer: Add a step to verify the deployment.

> 5. Run ``terraform apply``. After this command finishes, it should
>    output the URL to FLINT Cloud (ingress).

Reviewer 2: ```suggestion
5. Run ``terraform apply``. 

    After this command finishes, it outputs the URL to FLINT Cloud (ingress).
```
It's clearer for the reader if you put an expected result of completing an instruction on a separate line.

> To teardown the infrastructure and delete the application, run
> ``terraform destroy`` in the same directory as where ``main.tf`` is
> present. In case this fails, simply run it again.

Reviewer 2: ```suggestion
To tear down the infrastructure and delete the application, run
``terraform destroy`` in the same directory where ``main.tf`` is
present. If this fails, run it again.
```

.. _Terraform: https://www.terraform.io/

Azure
-----

> To deploy FLINT.Cloud to Azure, we recommend using the Azure App Service
> using a custom Microsoft Container Registry container, built from our
> ``local`` directory. This setup includes the usage of `Azure CLI`_ to
> deploy the infrastructure and the application to be used. You will also
> need to sign in to the Azure CLI by using the ``az login`` command. To
> finish the authentication process, follow the steps displayed in your
> terminal. To build the images, we would be using `Docker`_ and would be
> pushing them to `Azure Container Registry`_.

Reviewer 2: ```suggestion
To deploy FLINT.Cloud to Azure, we recommend using the Azure App Service
with a custom Microsoft Container Registry container built from our
``local`` directory. This setup uses `Azure CLI`_ to
deploy the infrastructure and the applications to be used. You need to 
sign in to the Azure CLI by using the ``az login`` command. To
finish the authentication process, follow the steps displayed in your
terminal. To build images, we use `Docker`_ and then push them to `Azure Container Registry`_.
```

Download the project
~~~~~~~~~~~~~~~~~~~~

Clone the repository:

.. code:: sh

   git clone https://github.com/moja-global/flint.cloud

Navigate to the ``local`` directory:

.. code:: sh

   sh local

Build the images locally
~~~~~~~~~~~~~~~~~~~~~~~~

To build the ``rest_api_gcbm`` image locally, run the following command:

.. code:: sh

   pushd rest_api_gcbm
   docker build --build-arg BUILD_TYPE=RELEASE --build-arg NUM_CPU=4 -t gcbm-api .
   popd

To build the ``rest_api_flint.example`` image locally, run the following
command:

.. code:: sh

   pushd rest_api_flint.example
   docker build -t flint-api .
   popd

Create a resource group
~~~~~~~~~~~~~~~~~~~~~~~

> You can now prepare resources in Azure to which you can push the image
> and then deploy a container to Azure App Service. You can start by
> creating a resource group in which you want to collect all the
> resources.

Reviewer 2: ```suggestion
To push images onto and deploy containers with the Azure App 
Service, you need to first prepare some resources. Start by
creating a resource group that will collect all your
resources.
```

.. code:: sh

   az group create --name myResourceGroup --location centralindia

You can change the ``--location`` value to specify a region near you.

Create a Container Registry
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can now push the image to Azure Container Registry from which App

Reviewer 2: ```suggestion
You can now push the image to Azure Container Registry so that App
```

Service can deploy it. Create an Azure Container Registry to push your
images to:

.. code:: sh

   az acr create --name <registry-name> --resource-group myResourceGroup --sku Basic --admin-enabled true

> Replace ``<registry-name>`` with a suitable name for your registry. The
> name must contain only letters, numbers, and must be unique across all

Reviewer 2: ```suggestion
Replace ``<registry-name>`` with a suitable name for your registry. The
name must contain only letters and numbers, and must be unique across all
```

of Azure.

Retrieve your credentials for the Container Registry:

.. code:: sh

   az acr credential show --resource-group myResourceGroup --name <registry-name>

Use the ``docker login`` command to sign in to the container registry:

.. code:: sh

   docker login <registry-name>.azurecr.io --username <registry-username>

Replace ``<registry-name>`` and ``<registry-username>`` with values from
the previous steps. When prompted, type in one of the passwords from the
previous step.

> Let's tag the images with the registry name:

Reviewer 2: ```suggestion
Tag the images with the registry name:
```
Google's style guide recommends to avoid this kind of phrase: https://developers.google.com/style/tone?hl=en#some-things-to-avoid-where-possible (bullet 12).

.. code:: sh

   docker tag rest_api_gcbm <registry-name>.azurecr.io/rest_api_gcbm:latest
   docker tag rest_api_flint.example <registry-name>.azurecr.io/rest_api_flint.example:latest

Use the docker push command to push the image to the registry:

.. code:: sh

   docker push <registry-name>.azurecr.io/rest_api_gcbm:latest
   docker push <registry-name>.azurecr.io/rest_api_flint.example:latest

Use the ``az acr repository list`` command to verify that the push was
successful:

.. code:: sh

   az acr repository list -n <registry-name>

Deploy the image from registry
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To deploy a container to Azure App Service, you first create a web app
on App Service, then connect the web app to the container registry. When
the web app starts, App Service automatically pulls the image from the
registry.

Create an App Service plan using the ``az appservice plan create``
command:

.. code:: sh

   az appservice plan create --name myAppServicePlan --resource-group myResourceGroup --is-linux

> Create the web app with the ``az webpp create`` command. Since we are
> deploying two images to two different web apps, you would need to
> specify the commands twice. To deploy, ``rest_api_gcbm`` to the first

Reviewer 2: ```suggestion
Create the web app with the ``az webpp create`` command. Since we are
deploying two images to two different web apps, you need to
enter these commands twice. To deploy ``rest_api_gcbm`` to the first
```

web app and ``rest_api_flint.example`` to the second web app, run the
following commands:

.. code:: sh

   az webapp create --resource-group myResourceGroup --plan myAppServicePlan --name <app-name-1> --deployment-container-image-name <registry-name>.azurecr.io/rest_api_gcbm:latest
   az webapp create --resource-group myResourceGroup --plan myAppServicePlan --name <app-name-2> --deployment-container-image-name <registry-name>.azurecr.io/rest_api_flint.example:latest

> Use the ``az webapp config appsettings set`` to set the
> ``WEBSITES_PORT`` environment variable as expected by the app code. In

Reviewer 2: ```suggestion
Use the ``az webapp config appsettings set`` to set the
``WEBSITES_PORT`` environment variable. In
```

   az webapp identity assign --resource-group myResourceGroup --name <app-name-2> --query principalId --output tsv

Replace ``<app-name>`` with the name you used in the previous step. The
output of the command (filtered by the ``--query`` and ``--output``
arguments) is the service principal of the assigned identity.

Retrieve your subscription ID with the ``az account show`` command,
which you need in the next step:

.. code:: sh

   az account show --query id --output tsv

Grant the managed identity permission to access the container registry:

.. code:: sh

   az role assignment create --assignee <principal-id> --scope /subscriptions/<subscription-id>/resourceGroups/myResourceGroup/providers/Microsoft.ContainerRegistry/registries/<registry-name> --role "AcrPull"

Replace the following values:

-  ``<principal-id>`` with the service principal ID from the
   ``az webapp identity assign`` command.
-  ``<registry-name>`` with the name of your container registry.
-  ``<subscription-id>`` with the subscription ID retrieved from the
   ``az account show`` command.

Make sure the above steps are repeated for both of the apps that you are

> doing to deploy. Configure your app to use the managed identity to pull

Reviewer 2: ```suggestion
going to deploy. Configure your app to use the managed identity to pull
```

   az resource update --ids /subscriptions/<subscription-id>/resourceGroups/myResourceGroup/providers/Microsoft.Web/sites/<app-name-2>/config/web --set properties.acrUseManagedIdentityCreds=True

Replace the following values:

-  ``<subscription-id>`` with the subscription ID retrieved from the az
   account show command.
-  ``<app-name>`` with the name of your web app.

Deploy the image
~~~~~~~~~~~~~~~~

Use the ``az webapp config container set`` command to specify the
container registry and the image to deploy for the web app:

.. code:: sh

   az webapp config container set --name <app-name-1> --resource-group myResourceGroup --docker-custom-image-name <registry-name>.azurecr.io/rest_api_gcbm:latest --docker-registry-server-url https://<registry-name>.azurecr.io
   az webapp config container set --name <app-name-2> --resource-group myResourceGroup --docker-custom-image-name <registry-name>.azurecr.io/rest_api_flint.example:latest --docker-registry-server-url https://<registry-name>.azurecr.io

> Replace ``<app-name-1>`` and ``<app-name-2>`` with the name of your web
> app and replace ``<registry-name>`` in two places with the name of your
> registry. When the ``az webapp config container set`` command completes,
> the web app should be running in the container on App Service.

Reviewer 2: ```suggestion
Replace ``<app-name-1>`` and ``<app-name-2>`` with the name of your web
app, and replace all instances of ``<registry-name>`` with the name of your
registry. When the ``az webapp config container set`` command completes,
the web app is running in the container on App Service.
```
